numpy/_core/src/multiarray/item_selection.c

```diff
@@ -303,7 +303,7 @@ PyArray_TakeFrom(PyArrayObject *self0, PyObject *indices0, int axis,
              */
             flags |= NPY_ARRAY_ENSURECOPY;
         }
-        dtype = PyArray_DESCR(self);
+        dtype = PyArray_DESCR(out);
```

Comment:
Tests are failing, but that isn't surprising.  The dtype must be `self`, because this ensures that the core operation happens with a single identitcal dtype (the input arrays, at the cost of a copy if needed).

In either case, this sis just not right.  What is probably required is to add force-casting to the flags `NPY_FORCECAST`(?).  But additionally add an explicit cast safety check.

TBH, there should be many other functions that implement `out=` which get it right.

Reply:
If you can point me to a function from where to steal that would help me a lot.

Reply:
Not sure how consistent we are.  There are two possibilities:
1. You allow same-kind casts (probably), but make sure the direction of the cast check is correct.
2. You only allow equivalent/no casts.

I don't care much, probably 1. is just as well, we tend to copy anyway often enough, and frankly, you could probably integrate the casting intot he core loop in the future in principle.

I think this should be OK as a (admittedly a bit rough) example: https://github.com/numpy/numpy/blob/main/numpy/_core/src/multiarray/multiarraymodule.c#L592

It doesn't use writeback if copy, but that logic should be fine (except forcing the cast).

Reply:
After seeing that take is used internally in a few more "ufunc like" operations, and those use same-kind casting (actually we can also use `NPY_DEFAULT_ASSIGN_CASTING` or so in C:
I would lean with going with that same-kind casting here.

---

This appears to not just fix take, but a chain of other operations.  Might be worthwhile to try to finish it.
The example I gave actually gives a `DeprecationWarning` which would make sense (unlike ufuncs, etc. we are not planning on `casting=` to allow a very simple resolution for all the places this kicks in.)

But for 2.0, I can also see just doing away with this:  It is highly untypical for NumPy in other places.

Reply:
What is your exact proposal? What should be done?

Reply:
I would go with testing for same-kind casting of the output, and otherwise give a DeprecationWarning, since it isn't a big hassle to do that.